src/Data/Json/JTable/Internal.purs

```diff
@@ -0,0 +1,218 @@
+module Data.Json.JTable.Internal
+  ( renderJTableRaw, renderRows, renderThead, renderTbody, tsToRows, sortTree
+  , Tree(..), tPath, tWidth, tHeight, tKids
+  , Cell(..), cCursor, cWidth, cHeight, cJsonPrim
+  , JPath(..), Table(..)
+  , tFromJson, tMergeArray, widthOfPrimTuple, padTree
+  , cFromJson, cMergeObj, mergeObjTuple
+  , foldJsonP, toPrim, zipWithIndex, orelse, strcmp, _nattr, _cspan, _rspan
+  ) where
+
+import Math (max)
+import Data.Either
+import Data.Maybe (Maybe(..), fromMaybe, isJust)
+import Data.Tuple
+import Data.String (joinWith, localeCompare)
+import Data.Array
+import Data.Foldable
+import qualified Data.StrMap as M
+import Data.Foldable (foldl, any, all, mconcat, elem)
+import Data.Traversable (for)
+import Control.Alt
+import Control.Apply
+import Control.Bind
+import Control.Monad
+import Control.MonadPlus
+
+import Data.Argonaut.Core
+import Data.Argonaut.JCursor
+import Data.Argonaut.Parser (jsonParser)
+import Text.Smolder.HTML (thead, tbody, tr, th, td)
+import Text.Smolder.Markup (Markup(..), Attribute(..), attribute, (!))
+
+-- path of object keys, with array indices omitted
+type JPath = [String]
+
+-- rows of cells
+type Table = [[Cell]]
+
+-- header data
+data Tree = T String JPath Number Number [Tree]
```

Comment:
For `JPath` elimination, my idea was to create a function similar to this:

```purescript
foldDownTree :: forall z. (Semigroup z) => (Tree -> JPath -> z) -> z
```

So your folding function has the current node, the path to the node (assembled using the labels in a top-down fashion), and can return anything as long as it can be merged with the final result (most likely `Table`??? Or some intermediate, mergeable structure?).

For the width and height, it's possible something simpler would suffice, which is just a recursive computation of the width and height (possibly memoized using `purescript-memoize` to avoid recomputation).

The reason I'm harping on this so much is because usually, these sorts of changes that eliminate duplication of information can have a *big* effect on the readability of code, its defect rate, and the ease with which it can be maintained over time in response to new feature enhancements. What are your thoughts on the above approach?